sparsecoding/inference.py

```diff
@@ -269,4 +269,112 @@ def infer(self, data, dictionary):
             
             self.checknan(a,'coefficients')
         return a
+    
+    
+    
+class LSM(InferenceMethod):
+    """
+    Infer coefficients for each image in data using elements dictionary.
+    Method implemented according to "Group Sparse Coding with a Laplacian Scale Mixture Prior" (P. J. Garrigues & B. A. Olshausen, 2010)
+    """    
+
+    def __init__(self, n_iter = 100, n_iter_LSM = 6, BETA=0.01, ALPHA=80.0, sigma=0.005, sparseThreshold = 10**-2, solver = None):
+        '''
+        
+        Parameters
+        ----------
+        n_iter : scalar (1,) default=100
+            number of iterations to run for an optimizer
+
+        n_iter_LSM : scalar (1,) default=6
+            number of iterations to run the outer loop of  LSM
+
+        BETA : scalar (1,) default=0.01
+            LSM parameter used to update LAMBDA          
+
+        ALPHA : scalar (1,) default=80.0
+            LSM parameter used to update LAMBDA  
+
+        sigma : scalar (1,) default=0.005
+            LSM parameter used to compute the loss function
+
+        sparseThreshold : scalar (1,) default=10**-2
+            threshold used to discard smallest coefficients in the final solution
+            SM parameter used to compute the loss function
+            
+        solver : default=None
+        '''
+        super().__init__(solver)
+        self.n_iter = n_iter
+        self.n_iter_LSM = n_iter_LSM
+        
+        self.BETA = BETA
+        self.ALPHA = ALPHA
+        self.sigma = sigma
+        self.sparseThreshold = sparseThreshold
+
+    def LSMLoss(self, Phi, x, s, Lambda, sigma):
+        
+        loss = (1/(2*(sigma**2)))*torch.pow(torch.norm(x - torch.mm(Phi,s.t()).t(), p=2, dim=1, keepdim=True),2) + torch.sum(Lambda.mul(torch.abs(s)), 1, keepdim=True)  
+        
+        return loss        
+     
+    def infer(self, data, dictionary):
+        """
+        Infer coefficients for each image in data using dict elements dictionary using Laplacian Scale Mixture (LSM)
+
+        Parameters
+        ----------
+        data : array-like (batch_size, n_features)
+            
+        dictionary : array-like, (n_features, n_basis)
+       
+        Returns
+        -------
+        coefficients : array-like (batch_size, n_basis)
+        """
+        batch_size, n_features = data.shape
+        n_features, n_basis = dictionary.shape
+        device = dictionary.device
+
+        # initialize
+        #u = torch.zeros((batch_size, n_basis)).to(device)
+        coefficients=torch.zeros(batch_size, n_basis, requires_grad=True).to(device)
+
+        for i in range(0,self.n_iter_LSM):
+        
+            Lambda = (self.ALPHA + 1)/(self.BETA + torch.abs(coefficients))
```

Comment:
Same comment on variable names here. 

I'm okay with python variable names having upper case parts, e.g. I and Phi...
`lambda` is a special function in python, so maybe call it something else or keep the capital letter at the start

Reply:
Used lambdas instead